crates/node/src/cluster_state_refresher/mod.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::{collections::BTreeMap, sync::Arc, time::Duration};
+
+use futures::StreamExt;
+use tokio::{
+    sync::watch,
+    time::{self, Instant, Interval, MissedTickBehavior},
+};
+use tracing::{debug, instrument, trace};
+
+use restate_core::{
+    cancellation_watcher,
+    network::{
+        Incoming, MessageRouterBuilder, MessageStream, NetworkSender, Networking, Outgoing,
+        TransportConnect,
+    },
+    Metadata, ShutdownError, TaskCenter, TaskKind,
+};
+use restate_types::{
+    cluster::cluster_state::{AliveNode, ClusterState, DeadNode, NodeState, SuspectNode},
+    config::{CommonOptions, Configuration},
+    net::cluster_state::{NodePing, NodePong},
+    time::MillisSinceEpoch,
+    GenerationalNodeId, PlainNodeId,
+};
+
+//todo(azmy): make these configurable
+const SUSPECT_THRESHOLD_FACTOR: u32 = 2;
+const DEAD_THRESHOLD_FACTOR: u32 = 4;
+
+#[derive(Clone)]
+struct SeenState {
+    generational_node_id: GenerationalNodeId,
+    seen_at: MillisSinceEpoch,
+}
+
+impl SeenState {
+    fn new(generational_node_id: GenerationalNodeId) -> Self {
+        SeenState {
+            generational_node_id,
+            seen_at: MillisSinceEpoch::now(),
+        }
+    }
+}
+
+#[derive(Clone, Default)]
+struct NodeTracker {
+    seen: Option<SeenState>,
+    last_attempt_at: Option<MillisSinceEpoch>,
+}
+
+pub struct ClusterStateRefresher<T> {
+    metadata: Metadata,
+    ping_requests: MessageStream<NodePing>,
+    ping_responses: MessageStream<NodePong>,
+    networking: Networking<T>,
+    nodes: BTreeMap<PlainNodeId, NodeTracker>,
+    heartbeat_interval: Duration,
+    cluster_state_watch_tx: watch::Sender<Arc<ClusterState>>,
```

Comment:
It's actually pretty common pattern here to wait for a certain state of the entire cluster or a single node. While this can be accomplished by waiting on changes of the cluster state in a state machine, we maybe can make the experience a little better by providing a wrapper on top of the watch that make it easier to wait on more complex configuration

Reply:
Do you have an example use-case?